### PR TITLE
Make PollingFileSystemWatcher serializable

### PR DESCRIPTION
**Pull request by MisinformedDNA**
Part of #2222

### CODE REVIEW

**Comment by MisinformedDNA** [comment]
@ahsonkhan @JeremyKuhne Can we get this reviewed?

File: src/System.IO.FileSystem.Watcher.Polling/System/IO/FileState.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+
+namespace System.IO
+{
+    [Serializable]
+    struct FileState
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher.Polling/System/IO/FileState.cs]: Accessibility modifier

File: src/System.IO.FileSystem.Watcher.Polling/System/IO/PathToFileStateHashtable.cs [hunk 2]
```diff
@@ -4,15 +4,15 @@
...
 
-namespace System.IO.FileSystem
+namespace System.IO
 {
     // this is a quick an dirty hashtable optimized for the PollingFileSystemWatcher
     // It allows mutating struct values (FileState) contained in the hashtable
     // It has optimized Equals and GetHasCode
     // It implements removals by marking values as "removed" (Path==null) and then garbage collecting them when table is resized
+    [Serializable]
     class PathToFileStateHashtable
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher.Polling/System/IO/PathToFileStateHashtable.cs]: Please fix the accessibility modifiers when you're changing these files

File: src/System.IO.FileSystem.Watcher.Polling/System/IO/PathToFileStateHashtable.cs [hunk 3]
```diff
@@ -138,7 +132,7 @@ private int ComputeBucket(ReadOnlySpan<char> file)
         private void Resize()
         {
             // this is because sometimes we just need to garbade collect instead of increase size
-            var newSize = Math.Max(_count * 2, 4);
+            var newSize = Math.Max(Count * 2, 4);
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher.Polling/System/IO/PathToFileStateHashtable.cs]: Also try to remove var when you're making changes. Var is ok for tests, but generally not for code.

**Comment by JeremyKuhne** [comment]
@KrzysztofCwalina, why did you want to make the watcher serializable?

File: src/System.IO.FileSystem.Watcher.Polling/System/IO/FileState.cs (hunk 1 above)
Comment (MisinformedDNA) [reply src/System.IO.FileSystem.Watcher.Polling/System/IO/FileState.cs]: Is there a reason we don't have a solution-wide analyzer to remind us about style guidelines?

**Comment by MisinformedDNA** [comment]
@JeremyKuhne @KrzysztofCwalina 

Part of the purpose of `PollingFileSystemWatcher` is that file events are too important to miss. If that is the case, then we need to account for file changes during downtime. I initially proposed import/export functions, but I was guided to aim for the more consistent `ISerializable`. Whatever direction we go, I believe we need a story for "file changes during downtime".

File: src/System.IO.FileSystem.Watcher.Polling/System/IO/FileState.cs (hunk 1 above)
Comment (JeremyKuhne) [reply src/System.IO.FileSystem.Watcher.Polling/System/IO/FileState.cs]: > Is there a reason we don't have a solution-wide analyzer to remind us about style guidelines?

You do get the feedback in VS, but it isn't super in your face. I think that was intentional.

image

In any case I'll bring it up with the team.

**Comment by JeremyKuhne** [comment]
> Part of the purpose of PollingFileSystemWatcher is that file events are too important to miss. If that is the case, then we need to account for file changes during downtime.

@MisinformedDNA Ah, I forgot that you were trying to do that. Sorry.

File: src/System.IO.FileSystem.Watcher.Polling/System/IO/PollingFileSystemWatcher.cs [hunk 4]
```diff
@@ -47,17 +51,44 @@ public PollingFileSystemWatcher(string path, string filter = "*", EnumerationOpt
...
             if (_disposed) throw new ObjectDisposedException(nameof(PollingFileSystemWatcher));
+            if (_started) return;
 
+            _started = true;
             ComputeChangesAndUpdateState(); // captures the initial state
-            _timer.Change(PollingIntervalInMilliseconds, Timeout.Infinite);
+            _timer.Change(PollingInterval, Timeout.Infinite);
+        }
+
+        private void AttemptRestart()
```
Comment (JeremyKuhne) [inline src/System.IO.FileSystem.Watcher.Polling/System/IO/PollingFileSystemWatcher.cs]: This should just be inline as there appears to be only one caller.

**Pull request merged**